src/ape/api/query.py

```diff
@@ -0,0 +1,107 @@
+import sys
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+from pydantic import BaseModel, PositiveInt, root_validator
+
+from ape.managers.networks import NetworkManager
+from ape.types import AddressType
+from ape.utils import abstractdataclass, abstractmethod
+
+# We can remove this once we stop supporting python3.7.
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+QueryType = Union["BlockQuery", "AccountQuery", "ContractEventQuery", "ContractMethodQuery"]
+
+
+class _BaseQuery(BaseModel):
+    type: str  # Used as discriminator
+    columns: Union[str, List[str]]
+
+
+class _BaseBlockQuery(_BaseQuery):
+    start_block: PositiveInt = 0
+    stop_block: PositiveInt
+
+    @root_validator(pre=True)
+    def check_start_block_before_stop_block(cls, values):
+        if values["stop_block"] < values["start_block"]:
+            raise ValueError(
+                f"stop_block: {values['stop_block']} cannot be less than "
+                f"start_block: {values['start_block']}."
+            )
+
+        return values
+
+
+class BlockQuery(_BaseBlockQuery):
+    type: Literal["blocks"]
+
+
+class _BaseAccountQuery(BaseModel):
+    start_nonce: PositiveInt = 0
+    stop_nonce: PositiveInt
+
+    @root_validator(pre=True)
+    def check_start_nonce_before_stop_nonce(cls, values):
+        if values["stop_nonce"] < values["start_nonce"]:
+            raise ValueError(
+                f"stop_nonce: {values['stop_nonce']} cannot be less than "
+                f"start_nonce: {values['start_nonce']}."
+            )
+
+        return values
+
+
+class AccountQuery(_BaseAccountQuery):
+    type: Literal["accounts"]
+    account: AddressType
+
+
+class ContractEventQuery(_BaseBlockQuery):
+    type: Literal["contract_events"]
+    contract: AddressType
+    event_id: bytes
+
+
+class ContractMethodQuery(_BaseBlockQuery):
+    type: Literal["contract_calls"]
+    contract: AddressType
+    method_id: bytes
```

Comment:
Wondering if it would be better to have a full `MethodABI` object here instead, so it can do validation of `method_args` as well as encoding :thinking: